MIR-borrowck: identify cause of (and fix) extra "move out"/"assign to" errors signalled by mir-borrowck

There are many tests in `compile-fail/borrowck/` that, in addition to signalling the errors that we already expect, *also* include a number of additional errors usually of the form "cannot move out of <path> because it is borrowed.".

 * You can see list of such tests in the [discrepancy spreadsheet](https://docs.google.com/spreadsheets/d/11RfQxTBpdcQW8y7S2ASHv6Wkljj9j2kQ81eJNxTl3KE/edit?usp=sharing).

 * Here is example output from an especially bad instance of the problem:

```
error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:22:2
   |
22 | }
   |  ^

error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:22:2
   |
22 | }
   |  ^

error[E0506]: cannot assign to `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:22:2
   |
22 | }
   |  ^

error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:32:2
   |
32 | }
   |  ^

error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:32:2
   |
32 | }
   |  ^

error[E0506]: cannot assign to `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:32:2
   |
32 | }
   |  ^

error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:42:2
   |
42 | }
   |  ^

error[E0505]: cannot move out of `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:42:2
   |
42 | }
   |  ^

error[E0506]: cannot assign to `vec` because it is borrowed (Mir)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-element-loan.rs:42:2
   |
42 | }
   |  ^
```

These errors often (perhaps always?) occur at the end of some lexical scope, so perhaps they are arising due to over-zealous handling of destructors, or `StorageDead` (or `EndRegion`)?

In any case, they represent a lot of extra noise and need to be dealt with.

I looked a little bit into this tonight, for the specific case of `borrowck-unary-move.rs`, since that is a pretty small self-contained example.

Here is the MIR:
<details>

```
// MIR for `foo`
// source = Fn(NodeId(4))
// pass_name = NLL
// disambiguator = after

fn foo(_1: std::boxed::Box<isize>) -> isize {
    let mut _0: isize;                   // return pointer
    scope 1 {
        let _2: std::boxed::Box<isize>;  // "x" in scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:8: 12:9
        scope 2 {
            let _3: &'13_0rs isize;      // "y" in scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:13:9: 13:10
        }
        scope 3 {
        }
    }
    let mut _4: ();
    let mut _5: std::boxed::Box<isize>;
    let mut _6: isize;

    bb0: {
        StorageLive(_2);                 // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:8: 12:9
        _2 = _1;                         // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:8: 12:9
        StorageLive(_3);                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:13:9: 13:10
        _3 = &'13_0rs (*_2);             // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:13:13: 13:16
        StorageLive(_5);                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:10: 14:11
        _5 = _2;                         // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:10: 14:11
        _4 = const free(_5) -> [return: bb1, unwind: bb6]; // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:5: 14:12
    }

    bb1: {
        drop(_5) -> [return: bb7, unwind: bb4]; // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:12: 14:12
    }

    bb2: {                               // cleanup
        resume;                          // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:1: 16:2
    }

    bb3: {                               // cleanup
        drop(_2) -> bb2;                 // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
    }

    bb4: {                               // cleanup
        drop(_1) -> bb3;                 // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
    }

    bb5: {                               // cleanup
        EndRegion('13_0rs);              // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:32: 16:2
        goto -> bb4;                     // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:32: 16:2
    }

    bb6: {                               // cleanup
        drop(_5) -> bb5;                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:12: 14:12
    }

    bb7: {
        StorageDead(_5);                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:14:12: 14:12
        StorageLive(_6);                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:15:5: 15:7
        _6 = (*_3);                      // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:15:5: 15:7
        _0 = _6;                         // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:15:5: 15:7
        StorageDead(_6);                 // scope 2 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:15:7: 15:7
        EndRegion('13_0rs);              // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:12:32: 16:2
        StorageDead(_3);                 // scope 1 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
        drop(_1) -> [return: bb8, unwind: bb3]; // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
    }

    bb8: {
        drop(_2) -> bb9;                 // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
    }

    bb9: {
        StorageDead(_2);                 // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
        return;                          // scope 0 at ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2
    }
}
```

</details>

and here is the line from RUST_LOG (with some line breaks added by hand for presentation purposes), right before we emit the extraneous error, along with the error itself:

```
DEBUG:rustc_mir::borrow_check: MirBorrowckCtxt::process_terminator(bb3[0], \
    Terminator { source_info: SourceInfo { span: ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2: 16:2, scope: scope0 }, \
                                           kind: drop(_2) -> bb2 }): \
    borrows in effect: [&'13_0rs (*_2)] borrows generated: [] inits: [_0, _3, _4, _6] uninits: [_0, _1, _2, _4, _5, _6]
error[E0505]: cannot move out of `x` because it is borrowed (Mir)
  --> ../src/test/compile-fail/borrowck/borrowck-unary-move.rs:16:2
   |
13 |     let y = &*x;
   |             --- borrow of `(*x)` occurs here
...
16 | }
   |  ^ move out of `x` occurs here
```

Two things of interest. 
 1. `_2` is not in the `inits` set, so we know from the dataflow that it cannot be initialized, which means this drop is a no-op. (Could/should we avoid emitting diagnostics for such case?)
 2. The borrow `&'13_0rs (*_2)` is apparently still in effect for *this* drop of `_2` (which is the sole content of `bb3`). There are three other drops of `_2` in the MIR, and the RUST_LOG output indicates that that borrow is not effect for any of those; just this particular one. So my immediate question is: Shouldn't there be an `EndRegion('13_0rs)` along the control-flow path before we get to bb3?

I think point 2 is the bigger bug of interest here. Figuring out what's happening there might resolve a lot of these extraneous E0505 emissions that I am seeing from MIR-borrowck.

--- a/src/librustc_mir/build/scope.rs
+++ b/src/librustc_mir/build/scope.rs
@@ -131,6 +131,9 @@ pub struct Scope<'tcx> {
 
     /// The cache for drop chain on "generator drop" exit.
     cached_generator_drop: Option<BasicBlock>,
+
+    /// The cache for drop chain on "unwind" exit.
+    cached_unwind: CachedBlock,
 }
 
 #[derive(Debug)]
@@ -233,8 +236,10 @@ impl<'tcx> Scope<'tcx> {
         self.cached_exits.clear();
 
         if !storage_only {
-            // the current generator drop ignores storage but refers to top-of-scope
+            // the current generator drop and unwind ignore
+            // storage but refer to top-of-scope
             self.cached_generator_drop = None;
+            self.cached_unwind.invalidate();
         }
 
         if !storage_only && !this_scope_only {
@@ -243,26 +248,6 @@ impl<'tcx> Scope<'tcx> {
                     cached_block.invalidate();
                 }
             }
-        }
-    }
-
-    /// Returns the cached entrypoint for diverging exit from this scope.
-    ///
-    /// Precondition: the caches must be fully filled (i.e. diverge_cleanup is called) in order for
-    /// this method to work correctly.
-    fn cached_block(&self, generator_drop: bool) -> Option<BasicBlock> {
-        let mut drops = self.drops.iter().rev().filter_map(|data| {
-            match data.kind {
-                DropKind::Value { cached_block } => {
-                    Some(cached_block.get(generator_drop))
-                }
-                DropKind::Storage => None
-            }
-        });
-        if let Some(cached_block) = drops.next() {
-            Some(cached_block.expect("drop cache is not filled"))
-        } else {
-            None
         }
     }
 
@@ -374,7 +359,8 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             needs_cleanup: false,
             drops: vec![],
             cached_generator_drop: None,
-            cached_exits: FxHashMap()
+            cached_exits: FxHashMap(),
+            cached_unwind: CachedBlock::default(),
         });
     }
 
@@ -500,15 +486,16 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                                    TerminatorKind::Goto { target: b });
                 b
             };
+
+            // End all regions for scopes out of which we are breaking.
+            self.cfg.push_end_region(self.hir.tcx(), block, src_info, scope.region_scope);
+
             unpack!(block = build_scope_drops(&mut self.cfg,
                                               scope,
                                               rest,
                                               block,
                                               self.arg_count,
                                               true));
-
-            // End all regions for scopes out of which we are breaking.
-            self.cfg.push_end_region(self.hir.tcx(), block, src_info, scope.region_scope);
         }
 
         self.cfg.terminate(block, src_info, TerminatorKind::GeneratorDrop);
@@ -841,23 +828,45 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
         let source_info = scope.source_info(drop_data.span);
         match drop_data.kind {
             DropKind::Value { .. } => {
-                // Try to find the next block with its cached block
-                // for us to diverge into in case the drop panics.
+                // Try to find the next block with its cached block for us to
+                // diverge into, either a previous block in this current scope or
+                // the top of the previous scope.
+                //
+                // If it wasn't for EndRegion, we could just chain all the DropData
+                // together and pick the first DropKind::Value. Please do that
+                // when we replace EndRegion with NLL.
                 let on_diverge = iter.clone().filter_map(|dd| {
                     match dd.kind {
                         DropKind::Value { cached_block } => Some(cached_block),
                         DropKind::Storage => None
                     }
-                }).map(|cached_block| {
-                    cached_block
-                        .get(generator_drop)
-                        .unwrap_or_else(|| span_bug!(drop_data.span, "cached block not present?"))
-                }).next();
-                // If there’s no `cached_block`s within current scope,
-                // we must look for one in the enclosing scope.
-                let on_diverge = on_diverge.or_else(|| {
-                    earlier_scopes.iter().rev().flat_map(|s| s.cached_block(generator_drop)).next()
+                }).next().or_else(|| {
+                    if earlier_scopes.iter().any(|scope| scope.needs_cleanup) {
+                        // If *any* scope requires cleanup code to be run,
+                        // we must use the cached unwind from the *topmost*
+                        // scope, to ensure all EndRegions from surrounding
+                        // scopes are executed before the drop code runs.
+                        Some(earlier_scopes.last().unwrap().cached_unwind)
+                    } else {
+                        // We don't need any further cleanup, so return None
+                        // to avoid creating a landing pad. We can skip
+                        // EndRegions because all local regions end anyway
+                        // when the function unwinds.
+                        //
+                        // This is an important optimization because LLVM is
+                        // terrible at optimizing landing pads. FIXME: I think
+                        // it would be cleaner and better to do this optimization
+                        // in SimplifyCfg instead of here.
+                        None
+                    }
                 });
+
+                let on_diverge = on_diverge.map(|cached_block| {
+                    cached_block.get(generator_drop).unwrap_or_else(|| {
+                        span_bug!(drop_data.span, "cached block not present?")
+                    })
+                });
+
                 let next = cfg.start_new_block();
                 cfg.terminate(block, source_info, TerminatorKind::Drop {
                     location: drop_data.location.clone(),
@@ -948,14 +957,21 @@ fn build_diverge_scope<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
         };
     }
 
-    // Finally, push the EndRegion block, used by mir-borrowck. (Block
-    // becomes trivial goto after pass that removes all EndRegions.)
-    {
-        let block = cfg.start_new_cleanup_block();
-        cfg.push_end_region(tcx, block, source_info(span), scope.region_scope);
-        cfg.terminate(block, source_info(span), TerminatorKind::Goto { target: target });
-        target = block
-    }
+    // Finally, push the EndRegion block, used by mir-borrowck, and set
+    // `cached_unwind` to point to it (Block becomes trivial goto after
+    // pass that removes all EndRegions).
+    target = {
+        let cached_block = scope.cached_unwind.ref_mut(generator_drop);
+        if let Some(cached_block) = *cached_block {
+            cached_block
+        } else {
+            let block = cfg.start_new_cleanup_block();
+            cfg.push_end_region(tcx, block, source_info(span), scope.region_scope);
+            cfg.terminate(block, source_info(span), TerminatorKind::Goto { target: target });
+            *cached_block = Some(block);
+            block
+        }
+    };
 
     debug!("build_diverge_scope({:?}, {:?}) = {:?}", scope, span, target);
 

--- a/src/librustc_mir/build/scope.rs
+++ b/src/librustc_mir/build/scope.rs
@@ -836,24 +836,22 @@ fn build_scope_drops<'tcx>(cfg: &mut CFG<'tcx>,
                            generator_drop: bool)
                            -> BlockAnd<()> {
     debug!("build_scope_drops({:?} -> {:?})", block, scope);
-    let mut iter = scope.drops.iter().rev().peekable();
+    let mut iter = scope.drops.iter().rev();
     while let Some(drop_data) = iter.next() {
         let source_info = scope.source_info(drop_data.span);
         match drop_data.kind {
             DropKind::Value { .. } => {
                 // Try to find the next block with its cached block
                 // for us to diverge into in case the drop panics.
-                let on_diverge = iter.peek().iter().filter_map(|dd| {
+                let on_diverge = iter.clone().filter_map(|dd| {
                     match dd.kind {
-                        DropKind::Value { cached_block } => {
-                            let result = cached_block.get(generator_drop);
-                            if result.is_none() {
-                                span_bug!(drop_data.span, "cached block not present?")
-                            }
-                            result
-                        },
+                        DropKind::Value { cached_block } => Some(cached_block),
                         DropKind::Storage => None
                     }
+                }).map(|cached_block| {
+                    cached_block
+                        .get(generator_drop)
+                        .unwrap_or_else(|| span_bug!(drop_data.span, "cached block not present?"))
                 }).next();
                 // If there’s no `cached_block`s within current scope,
                 // we must look for one in the enclosing scope.

--- a/src/librustc_mir/build/scope.rs
+++ b/src/librustc_mir/build/scope.rs
@@ -154,6 +154,11 @@ struct CachedBlock {
     unwind: Option<BasicBlock>,
 
     /// The cached block for unwinds during cleanups-on-generator-drop path
+    ///
+    /// This is split from the standard unwind path here to prevent drop
+    /// elaboration from creating drop flags that would have to be captured
+    /// by the generator. I'm not sure how important this optimization is,
+    /// but it is here.
     generator_drop: Option<BasicBlock>,
 }
 
@@ -217,13 +222,26 @@ impl<'tcx> Scope<'tcx> {
     /// Should always be run for all inner scopes when a drop is pushed into some scope enclosing a
     /// larger extent of code.
     ///
-    /// `unwind` controls whether caches for the unwind branch are also invalidated.
-    fn invalidate_cache(&mut self, unwind: bool) {
+    /// `storage_only` controls whether to invalidate only drop paths run `StorageDead`.
+    /// `this_scope_only` controls whether to invalidate only drop paths that refer to the current
+    /// top-of-scope (as opposed to dependent scopes).
+    fn invalidate_cache(&mut self, storage_only: bool, this_scope_only: bool) {
+        // FIXME: maybe do shared caching of `cached_exits` etc. to handle functions
+        // with lots of `try!`?
+
+        // cached exits drop storage and refer to the top-of-scope
         self.cached_exits.clear();
-        if !unwind { return; }
-        for dropdata in &mut self.drops {
-            if let DropKind::Value { ref mut cached_block } = dropdata.kind {
-                cached_block.invalidate();
+
+        if !storage_only {
+            // the current generator drop ignores storage but refers to top-of-scope
+            self.cached_generator_drop = None;
+        }
+
+        if !storage_only && !this_scope_only {
+            for dropdata in &mut self.drops {
+                if let DropKind::Value { ref mut cached_block } = dropdata.kind {
+                    cached_block.invalidate();
+                }
             }
         }
     }
@@ -672,8 +690,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             // invalidating caches of each scope visited. This way bare minimum of the
             // caches gets invalidated. i.e. if a new drop is added into the middle scope, the
             // cache of outer scpoe stays intact.
-            let invalidate_unwind = needs_drop && !this_scope;
-            scope.invalidate_cache(invalidate_unwind);
+            scope.invalidate_cache(!needs_drop, this_scope);
             if this_scope {
                 if let DropKind::Value { .. } = drop_kind {
                     scope.needs_cleanup = true;
@@ -941,6 +958,8 @@ fn build_diverge_scope<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
         cfg.terminate(block, source_info(span), TerminatorKind::Goto { target: target });
         target = block
     }
+
+    debug!("build_diverge_scope({:?}, {:?}) = {:?}", scope, span, target);
 
     target
 }

--- a/src/test/compile-fail/borrowck/borrowck-unary-move.rs
+++ b/src/test/compile-fail/borrowck/borrowck-unary-move.rs
@@ -8,10 +8,15 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+// ignore-tidy-linelength
+// revisions: ast mir
+//[mir]compile-flags: -Z emit-end-regions -Z borrowck-mir
 
 fn foo(x: Box<isize>) -> isize {
     let y = &*x;
-    free(x); //~ ERROR cannot move out of `x` because it is borrowed
+    free(x); //[ast]~ ERROR cannot move out of `x` because it is borrowed
+    //[mir]~^ ERROR cannot move out of `x` because it is borrowed (Ast)
+    //[mir]~| ERROR cannot move out of `x` because it is borrowed (Mir)
     *y
 }
 

--- a/src/test/run-pass/dynamic-drop.rs
+++ b/src/test/run-pass/dynamic-drop.rs
@@ -177,6 +177,17 @@ fn generator(a: &Allocator, run_count: usize) {
     }
 }
 
+fn mixed_drop_and_nondrop(a: &Allocator) {
+    // check that destructor panics handle drop
+    // and non-drop blocks in the same scope correctly.
+    //
+    // Surprisingly enough, this used to not work.
+    let (x, y, z);
+    x = a.alloc();
+    y = 5;
+    z = a.alloc();
+}
+
 #[allow(unreachable_code)]
 fn vec_unreachable(a: &Allocator) {
     let _x = vec![a.alloc(), a.alloc(), a.alloc(), return];
@@ -244,11 +255,12 @@ fn main() {
     run_test(|a| field_assignment(a, false));
     run_test(|a| field_assignment(a, true));
 
-    // FIXME: fix leaks on panics
-    run_test_nopanic(|a| generator(a, 0));
-    run_test_nopanic(|a| generator(a, 1));
-    run_test_nopanic(|a| generator(a, 2));
-    run_test_nopanic(|a| generator(a, 3));
+    run_test(|a| generator(a, 0));
+    run_test(|a| generator(a, 1));
+    run_test(|a| generator(a, 2));
+    run_test(|a| generator(a, 3));
+
+    run_test(|a| mixed_drop_and_nondrop(a));
 
     run_test_nopanic(|a| union1(a));
 }

--- a/src/test/run-pass/dynamic-drop.rs
+++ b/src/test/run-pass/dynamic-drop.rs
@@ -8,9 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-#![feature(untagged_unions)]
+#![feature(generators, generator_trait, untagged_unions)]
 
 use std::cell::{Cell, RefCell};
+use std::ops::Generator;
 use std::panic;
 use std::usize;
 
@@ -161,6 +162,21 @@ fn vec_simple(a: &Allocator) {
     let _x = vec![a.alloc(), a.alloc(), a.alloc(), a.alloc()];
 }
 
+fn generator(a: &Allocator, run_count: usize) {
+    assert!(run_count < 4);
+
+    let mut gen = || {
+        (a.alloc(),
+         yield a.alloc(),
+         a.alloc(),
+         yield a.alloc()
+         );
+    };
+    for _ in 0..run_count {
+        gen.resume();
+    }
+}
+
 #[allow(unreachable_code)]
 fn vec_unreachable(a: &Allocator) {
     let _x = vec![a.alloc(), a.alloc(), a.alloc(), return];
@@ -227,6 +243,12 @@ fn main() {
 
     run_test(|a| field_assignment(a, false));
     run_test(|a| field_assignment(a, true));
+
+    // FIXME: fix leaks on panics
+    run_test_nopanic(|a| generator(a, 0));
+    run_test_nopanic(|a| generator(a, 1));
+    run_test_nopanic(|a| generator(a, 2));
+    run_test_nopanic(|a| generator(a, 3));
 
     run_test_nopanic(|a| union1(a));
 }

